Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,55 +47,127 @@ void RebinToWorkspace::init() {
+    if (left->getSpectrum(i).x() != rght->getSpectrum(i).x()) {
+      return true;
+    }
+  }
+  // everything must be the same
+  return false;
+}
+} // namespace
+
 /**
  * Execute the algorithm
  */
 void RebinToWorkspace::exec() {
   // The input workspaces ...
   MatrixWorkspace_sptr toRebin = getProperty("WorkspaceToRebin");
   MatrixWorkspace_sptr toMatch = getProperty("WorkspaceToMatch");
-  bool PreserveEvents = getProperty("PreserveEvents");
-
-  // First we need to create the parameter vector from the workspace with which
-  // we are matching
-  std::vector<double> rb_params = createRebinParameters(toMatch);
-
-  IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
-  runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
-  runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
-  runRebin->setProperty("params", rb_params);
-  runRebin->setProperty("PreserveEvents", PreserveEvents);
-  runRebin->setProperty("IgnoreBinErrors", true);
-  runRebin->executeAsChildAlg();
-  progress(1);
-  MatrixWorkspace_sptr ws = runRebin->getProperty("OutputWorkspace");
-  setProperty("OutputWorkspace", ws);
+
+  if (needToRebin(toRebin, toMatch)) {
+    g_log.information("Rebinning");
+    this->rebin(toRebin, toMatch);
+  } else { // don't need to rebin
+    g_log.information(
+        "WorkspaceToRebin and WorkspaceToMatch already have matched binning");
+    auto outputWS = this->createOutputWorkspace(toRebin, 1.);
+    this->setProperty("OutputWorkspace", outputWS);
+  }
 }
 
-/**
- * Create the vector of rebin parameters
- * @param toMatch :: A shared pointer to the workspace with the desired binning
- * @returns :: A vector to hold the rebin parameters once they have been
- * calculated
- */
-std::vector<double> RebinToWorkspace::createRebinParameters(
-    Mantid::API::MatrixWorkspace_sptr toMatch) {
-  using namespace Mantid::API;
-
-  auto &matchXdata = toMatch->x(0);
-  // params vector should have the form [x_1, delta_1,x_2, ...
-  // ,x_n-1,delta_n-1,x_n), see Rebin.cpp
-  std::vector<double> rb_params;
-  int xsize = static_cast<int>(matchXdata.size());
-  rb_params.reserve(xsize * 2);
-  for (int i = 0; i < xsize; ++i) {
-    // bin bound
-    rb_params.push_back(matchXdata[i]);
-    // Bin width
-    if (i < xsize - 1)
-      rb_params.push_back(matchXdata[i + 1] - matchXdata[i]);
+MatrixWorkspace_sptr
+RebinToWorkspace::createOutputWorkspace(MatrixWorkspace_sptr &inputWS,
+                                        const double endProgress) {
+  MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+  bool preserveEvents = getProperty("PreserveEvents");
+
+  bool inPlace = (inputWS == outputWS);
+  bool isEvents =
+      bool(boost::dynamic_pointer_cast<const EventWorkspace>(inputWS));
+
+  if (inPlace) {
+    if (isEvents && (!preserveEvents)) {
+      // convert to a MatrixWorkspace
+      IAlgorithm_sptr convert =
+          createChildAlgorithm("ConvertToMatrixWorkspace");
```

Comment:
So by first converting to `Workspace2D` (I assume the algorithm name is wrong? `EventWorkspace` *is* a `MatrixWorkspace`) you are effectively rebinning twice, and potentially/likely lose information. For example, if the event workspaces only has the "default" two min/max bin boundaries, you just get a flat result?

I think there are at least two options to avoid this:
1. First set the new bin boundaries on the event workspace, then call `ConvertToMatrixWorkspace` (no need to do a rebin afterwards, but I think you are forced to temporarily copy the events).
2. Write `rebin` below to directly work with an input `EventWorkspace`, using something like `EventList::generateHistogram` with the desired target edges.